packages/grid/x-data-grid/src/tests/rows.DataGrid.test.tsx

```diff
@@ -414,68 +418,258 @@ describe('<DataGrid /> - Rows', () => {
+      });
 
-    it('should set the second row to have a different row height than the others', () => {
-      render(<TestCase getRowHeight={({ id }) => (id === 'c2' ? 100 : null)} />);
+      it('should set the second row to have a different row height than the others', () => {
+        render(<TestCase getRowHeight={({ id }) => (id === 'c2' ? 100 : null)} />);
 
-      expect(getRow(0).clientHeight).to.equal(ROW_HEIGHT);
-      expect(getRow(1).clientHeight).to.equal(100);
-      expect(getRow(2).clientHeight).to.equal(ROW_HEIGHT);
-    });
+        expect(getRow(0).clientHeight).to.equal(ROW_HEIGHT);
+        expect(getRow(1).clientHeight).to.equal(100);
+        expect(getRow(2).clientHeight).to.equal(ROW_HEIGHT);
+      });
 
-    it('should set density to all but the row with variable row height', () => {
-      const { setProps } = render(
-        <TestCase getRowHeight={({ id }) => (id === 'c2' ? 100 : null)} />,
-      );
+      it('should set density to all but the row with variable row height', () => {
+        const { setProps } = render(
+          <TestCase getRowHeight={({ id }) => (id === 'c2' ? 100 : null)} />,
+        );
+
+        expect(getRow(0).clientHeight).to.equal(ROW_HEIGHT);
+        expect(getRow(1).clientHeight).to.equal(100);
+        expect(getRow(2).clientHeight).to.equal(ROW_HEIGHT);
+
+        setProps({ density: 'compact' });
+
+        expect(getRow(0).clientHeight).to.equal(Math.floor(ROW_HEIGHT * COMPACT_DENSITY_FACTOR));
+        expect(getRow(1).clientHeight).to.equal(100);
+        expect(getRow(2).clientHeight).to.equal(Math.floor(ROW_HEIGHT * COMPACT_DENSITY_FACTOR));
+      });
+
+      it('should set the correct rowHeight and variable row height', () => {
+        const { setProps } = render(
+          <TestCase getRowHeight={({ id }) => (id === 'c2' ? 100 : null)} />,
+        );
 
-      expect(getRow(0).clientHeight).to.equal(ROW_HEIGHT);
-      expect(getRow(1).clientHeight).to.equal(100);
-      expect(getRow(2).clientHeight).to.equal(ROW_HEIGHT);
+        expect(getRow(0).clientHeight).to.equal(ROW_HEIGHT);
+        expect(getRow(1).clientHeight).to.equal(100);
+        expect(getRow(2).clientHeight).to.equal(ROW_HEIGHT);
 
-      setProps({ density: 'compact' });
+        setProps({ rowHeight: 30 });
 
-      expect(getRow(0).clientHeight).to.equal(Math.floor(ROW_HEIGHT * COMPACT_DENSITY_FACTOR));
-      expect(getRow(1).clientHeight).to.equal(100);
-      expect(getRow(2).clientHeight).to.equal(Math.floor(ROW_HEIGHT * COMPACT_DENSITY_FACTOR));
+        expect(getRow(0).clientHeight).to.equal(30);
+        expect(getRow(1).clientHeight).to.equal(100);
+        expect(getRow(2).clientHeight).to.equal(30);
+      });
     });
 
-    it('should set the correct rowHeight and variable row height', () => {
-      const { setProps } = render(
-        <TestCase getRowHeight={({ id }) => (id === 'c2' ? 100 : null)} />,
-      );
+    describe('dynamic row height', () => {
```

Comment:
Can we have a test that validates that the density works as expected when only specific rows have either a dynamic or variable row height? I don't remember adding a test like that when I did the variable row height feature.